tools/disasm/disasm.py

```diff
@@ -0,0 +1,136 @@
+    context: spimdisasm.common.Context,
+    config_dir: Path,
+    dma_file: DmaFile,
+    f: BinaryIO,
+) -> spimdisasm.mips.FileSplits:
+    # Assume that we're reading from a decompressed ROM where the DMA file is
+    # now located at the same ROM offset as the VROM start
+    f.seek(dma_file.vrom_start)
+    data = bytearray(f.read(dma_file.vrom_end - dma_file.vrom_start))
+
+    file_splits_path = config_dir / f"files_{dma_file.name}.csv"
+    if file_splits_path.exists():
+        default_filename = ""
+        splits_data = spimdisasm.common.FileSplitFormat()
+        splits_data.readCsvFile(file_splits_path)
+        reloc_section = None
+    elif dma_file.overlay_dir is not None:
+        z_name = get_z_name_for_overlay(dma_file.name)
+        default_filename = f"src/overlays/{dma_file.overlay_dir}/{dma_file.name}/{z_name}.s"
+        splits_data = None
+        reloc_section = spimdisasm.mips.sections.SectionRelocZ64(
+            context,
+            vromStart=0,
+            vromEnd=len(data),
+            vram=dma_file.vram_start,
+            filename=default_filename,
+            array_of_bytes=data,
+            segmentVromStart=0,
+            overlayCategory=None)
+    else:
+        raise Exception(f"DMA file {dma_file.name} is not an overlay but has no file splits")
+
+    return spimdisasm.mips.FileSplits(
+        context,
+        vromStart=0,
+        vromEnd=len(data),
+        vram=dma_file.vram_start,
+        filename=default_filename,
+        array_of_bytes=data,
+        segmentVromStart=0,
+        overlayCategory=None,
+        splitsData=splits_data,
+        relocSection=reloc_section)
+
+def main():
+    parser = argparse.ArgumentParser(description="Disassemble a ROM.")
+    parser.add_argument("rom", type=Path, help="Input ROM")
+    parser.add_argument("-o", "--output-dir", help="Output directory", type=Path, required=True)
+    parser.add_argument("--config-dir", help="Config directory", type=Path, required=True)
+    parser.add_argument("--split-functions", help="Write functions into separate files", type=Path)
+
+    spimdisasm.common.Context.addParametersToArgParse(parser)
+    spimdisasm.common.GlobalConfig.addParametersToArgParse(parser)
+    spimdisasm.mips.InstructionConfig.addParametersToArgParse(parser)
+
+    args = parser.parse_args()
+
+    context = spimdisasm.common.Context()
+    context.parseArgs(args)
+    context.changeGlobalSegmentRanges(0x00000000, 0x01000000, 0x8000000, 0x81000000)
+    context.addBannedSymbolRange(0x10000000, 0x80000300)
+    context.addBannedSymbolRange(0xA0000000, 0xFFFFFFFF)
+
+    spimdisasm.mips.InstructionConfig.parseArgs(args)
+    spimdisasm.common.GlobalConfig.parseArgs(args)
+
+    spimdisasm.common.GlobalConfig.ASM_USE_PRELUDE = False
+    spimdisasm.common.GlobalConfig.PRODUCE_SYMBOLS_PLUS_OFFSET = True
+    spimdisasm.common.GlobalConfig.TRUST_USER_FUNCTIONS = True
+
+    dma_files = parse_file_addresses(args.config_dir / "file_addresses.csv")
```

Comment:
I kinda think this will work correctly and disassembly 0 files. Maybe that's confusing, but `file_addresses.csv` is committed here so I don't think it's a danger

Reply:
yeah it should be fine, I was just thinking of the very unlikely situation where someone doesn't have that file for some reasons